### PR TITLE
Remove the dependency on dash.

### PR DESCRIPTION
**Pull request by zhenya1007**
A few simple mechanical transformations to remove the dependency on dash.

### CODE REVIEW

**Comment by zhenya1007** [comment]
Would you consider merging something like this?  It doesn't seem like using dash gains you a lot here.

**Comment by tarsius** [comment]
> Would you consider merging something like this? It doesn't seem like using dash gains you a lot here.

I try not to depend on `dash` in low-level libraries but for historic reasons `with-editor` uses it anyway (it used to be part of `magit`). Never-the-less, my initial response was to reject this proposal. However I have just made some changes into this direction and now I am undecided, leaning toward approving it.

I have rebased your commit onto my changes and pushed them to your branch. Please replace `--each` with `dolist` instead of `mapc` and name the variable `file` instead of `it`. Also rename the variable to `elt` instead of `it` in the two added `lambda`s. Now I am inclined to merge this because the cost is only two unnecessary lambdas (I really wished Emacs provided clojure-style anonymous lambdas).

**Comment by tarsius** [comment]
Also please remove the period at the end of the commit message summary.

File: with-editor.el [hunk 1]
```diff
@@ -662,10 +661,10 @@ else like the former."
 (defun with-editor-shell-command-read-args (prompt &optional async)
   (let ((command (read-shell-command
                   prompt nil nil
-                  (--when-let (or buffer-file-name
-                                  (and (eq major-mode 'dired-mode)
-                                       (dired-get-filename nil t)))
-                    (file-relative-name it)))))
+                  (when-let (file (or buffer-file-name
```
Comment (npostavs) [inline with-editor.el]: `when-let` is new in Emacs 25.1 (and we're requiring only 24.4 currently)

Comment (tarsius) [inline with-editor.el]: I thought it was new in `24.4`... reverting.

Comment (tarsius) [inline with-editor.el]: @zhenya1007 I am afraid that means that I am not willing to drop the dependency on `dash`, sorry.

Comment (zhenya1007) [inline with-editor.el]: Well, it was worth a try. :-)

**Pull request closed without merging**